data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -291,16 +291,15 @@ private static Schema fileProjection(
       return requestedSchema;
     }
 
-    // TODO: support adding nested columns. this will currently fail when finding nested columns to
-    // add
+    // Add missing columns required for delete operations, including nested columns
     List<Types.NestedField> columns = Lists.newArrayList(requestedSchema.columns());
     for (int fieldId : missingIds) {
       if (fieldId == MetadataColumns.ROW_POSITION.fieldId()
           || fieldId == MetadataColumns.IS_DELETED.fieldId()) {
         continue; // add _pos and _deleted at the end
       }
 
-      Types.NestedField field = tableSchema.asStruct().field(fieldId);
+      Types.NestedField field = getFieldFromTableSchema(tableSchema, fieldId);
```

Comment:
It looks incorrect to me to append a nested field directly to the top-level schema. For missing nested IDs, I think we need to preserve their full path, not promote the leaf to top level.